lib/util/source-code.js

```diff
@@ -138,7 +139,26 @@ function SourceCode(text, ast) {
      * This is done to avoid each rule needing to do so separately.
      * @type string[]
      */
-    this.lines = SourceCode.splitLines(this.text);
+    this.lines = [];
+    this.lineStartIndices = [0];
+
+    const lineEndingPattern = astUtils.createGlobalLinebreakMatcher();
+    let match;
+
+    /*
+     * Previously, this was implemented using a regex that
+     * matched a sequence of non-linebreak characters followed by a
+     * linebreak, then adding the lengths of the matches. However,
+     * this caused a catastrophic backtracking issue when the end
+     * of a file contained a large number of non-newline characters.
+     * To avoid this, the current implementation just matches newlines
+     * and uses match.index to get the correct line start indices.
+     */
+    while ((match = lineEndingPattern.exec(this.text))) {
```

Comment:
Yes, it was found in https://github.com/eslint/eslint/pull/7898.

I'm not sure I understand your performance comparison, since this is the same implementation as the previous `ast-utils` version. Could you clarify ?